src/auth/oauth/mod.rs

```diff
@@ -0,0 +1,446 @@
+use crate::auth::get_user_from_headers;
+use crate::auth::oauth::uris::{OAuthRedirectUris, ValidatedRedirectUri};
+use crate::auth::validate::extract_authorization_header;
+use crate::database::models::flow_item::Flow;
+use crate::database::models::oauth_client_authorization_item::OAuthClientAuthorization;
+use crate::database::models::oauth_client_item::OAuthClient as DBOAuthClient;
+use crate::database::models::oauth_token_item::OAuthAccessToken;
+use crate::database::models::{
+    generate_oauth_access_token_id, generate_oauth_client_authorization_id,
+    OAuthClientAuthorizationId, OAuthClientId,
+};
+use crate::database::redis::RedisPool;
+use crate::models;
+use crate::models::pats::Scopes;
+use crate::queue::session::AuthQueue;
+use actix_web::web::{scope, Data, Query, ServiceConfig};
+use actix_web::{get, post, web, HttpRequest, HttpResponse};
+use chrono::Duration;
+use rand::distributions::Alphanumeric;
+use rand::{Rng, SeedableRng};
+use rand_chacha::ChaCha20Rng;
+use reqwest::header::{CACHE_CONTROL, LOCATION, PRAGMA};
+use serde::{Deserialize, Serialize};
+use sqlx::postgres::PgPool;
+
+use self::errors::{OAuthError, OAuthErrorType};
+
+use super::AuthenticationError;
+
+pub mod errors;
+pub mod uris;
+
+pub fn config(cfg: &mut ServiceConfig) {
+    cfg.service(
+        scope("auth/oauth")
+            .service(init_oauth)
+            .service(accept_client_scopes)
+            .service(reject_client_scopes)
+            .service(request_token),
+    );
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct OAuthInit {
+    pub client_id: String,
+    pub redirect_uri: Option<String>,
+    pub scope: Option<String>,
+    pub state: Option<String>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct OAuthClientAccessRequest {
+    pub flow_id: String,
+    pub client_id: OAuthClientId,
+    pub client_name: String,
+    pub client_icon: Option<String>,
+    pub requested_scopes: Scopes,
+}
+
+#[get("authorize")]
+pub async fn init_oauth(
+    req: HttpRequest,
+    Query(oauth_info): Query<OAuthInit>,
+    pool: Data<PgPool>,
+    redis: Data<RedisPool>,
+    session_queue: Data<AuthQueue>,
+) -> Result<HttpResponse, OAuthError> {
+    let user = get_user_from_headers(
+        &req,
+        &**pool,
+        &redis,
+        &session_queue,
+        Some(&[Scopes::USER_AUTH_WRITE]),
+    )
+    .await
+    .map_err(OAuthError::error)?
```

Comment:
Yeah, not thrilled with how cumbersome the oauth error type turned out to be, but as outlined in my comment above, we need the error to take one of two forms - either a redirect with url-encoded error info or an error response directly back to the user agent.

So each time we're doing error handling in the oauth methods, we need to choose which of those is a better match - so we can't just let an auto-conversion happen using the `?` operator. It's critical to security that we don't redirect the user to the provided callback if the provided callback is invalid.

Reply:
could we set it up so that one of the two (the one that doesnt redirect to the provided callback) is ?-able? Or is this just ensuring we dont use ? no matter what the case so the programmer wont accidentally mix them up

Reply:
Hmm interesting yeah we could do that - basically have it default to OAuthError::error. That's the safer of the two options security-wise, so if we do that when we meant to redirect it's not the end of the world, the third party just won't get a callback telling them what happened. Plus, OAuthError::error is the one that needs no additional parameters, while OAuthError::redirect does.

I think there is some value in making sure we don't mix them up, but it's probably not worth the burden of mapping it all over the place. Thanks for the suggestion, I'll do that

Reply:
Alright we're using `?` now for the error situations, still need to be explicit for redirects